### ISSUE TITLE
After uploading files through kong, the time consumption increased obviously(request_time、upstream_response_time).

### ISSUE THREAD

**Issue by 1798972** [Kong/kong#10633]
## 1. Background

Now I am doing a file upload scene, the purpose is to upload pictures from the front page to the background server, this process through the kong transfer.
> Front App -> kong -> Background Service

Now after uploading through kong, the response time is long, which is the information in the log.

> kongDealTime: $request_time
>
> upRespTime: $upstream_response_time
```bash
 ===> 2023-04-06T18:55:51+08:00 | 576208ee-adaa-3bbe-fbfc-682bf32609ef | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [3.570] | upRespTime: [2.452] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [http://localhost:4000/] | reqClientInfo: [Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1] | respBytes:[685] | respBodyBytes:[275] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T18:56:15+08:00 | 9bf600ef-e70d-db00-a6ff-d86a14a23f52 | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [2.509] | upRespTime: [1.442] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [http://localhost:4000/] | reqClientInfo: [Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1] | respBytes:[685] | respBodyBytes:[275] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T18:56:45+08:00 | 441886fd-4e3b-600c-7376-2d0cbf263d65 | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [15.783] | upRespTime: [14.316] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [http://localhost:4000/] | reqClientInfo: [Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1] | respBytes:[1207] | respBodyBytes:[796] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T18:57:29+08:00 | e6f55acf-8f5a-c87c-d5d6-7dd5afb14315 | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [29.265] | upRespTime: [26.986] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [http://localhost:4000/] | reqClientInfo: [Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1] | respBytes:[476] | respBodyBytes:[66] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T18:57:45+08:00 | 9e29b939-9e30-98ac-70d0-c8ab33615245 | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [8.540] | upRespTime: [5.474] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [http://localhost:4000/] | reqClientInfo: [Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1] | respBytes:[685] | respBodyBytes:[275] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T18:58:25+08:00 | ed03dacf-264b-3183-25fd-8e1da73dcbab | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [32.343] | upRespTime: [23.665] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [http://localhost:4000/] | reqClientInfo: [Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1] | respBytes:[476] | respBodyBytes:[66] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T18:58:57+08:00 | 27c94f2f-9ba6-1576-8904-63cc622fcf0f | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [4.777] | upRespTime: [1.399] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [http://localhost:4000/] | reqClientInfo: [Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1] | respBytes:[685] | respBodyBytes:[275] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T18:59:35+08:00 | 545a4c10-cc64-f9f8-2140-ddcdc69277a8 | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [30.413] | upRespTime: [27.301] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [http://localhost:4000/] | reqClientInfo: [Mozilla/5.0 (iPhone; CPU iPhone OS 13_2_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.3 Mobile/15E148 Safari/604.1] | respBytes:[1205] | respBodyBytes:[794] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T19:21:34+08:00 | 89fbff12-8965-f8d4-c1dc-3b26edc479ab | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [3.638] | upRespTime: [2.388] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [-] | reqClientInfo: [PostmanRuntime/7.31.1] | respBytes:[685] | respBodyBytes:[275] | upAddr: 10.7.12.130:8889
 ===> 2023-04-06T19:42:03+08:00 | 70014d9f-5240-4735-d9cf-fd25fe3c1d07 | schemeVersion: -1 | messageType: 0.0.0.0.0 | reqUri: POST /wkg-h5-api/02/ANY/IBP/ibp/wkg/ocr/ocrIdCardAndUploadOss HTTP/1.1 | kongRespCoed: [200] | upRespCode: [200] | kongDealTime: [3.692] | upRespTime: [1.406] | reqFrom: 10.7.240.30 | forwardedFor: 10.16.88.180, 10.7.240.30 | urlRefer: [-] | reqClientInfo: [PostmanRuntime/7.31.1] | respBytes:[685] | respBodyBytes:[275] | upAddr: 10.7.12.130:8889
```
I'll simplify the above log file:
```bash
 $request_time: [3.570] | $upstream_response_time: [2.452] |
 $request_time: [2.509] | $upstream_response_time: [1.442] |
 $request_time: [15.783] | $upstream_response_time: [14.316]
 $request_time: [29.265] | $upstream_response_time: [26.986]
 $request_time: [8.540] | $upstream_response_time: [5.474] |
 $request_time: [32.343] | $upstream_response_time: [23.665]
 $request_time: [4.777] | $upstream_response_time: [1.399] |
 $request_time: [30.413] | $upstream_response_time: [27.301]
 $request_time: [3.638] | $upstream_response_time: [2.388] |
 $request_time: [3.692] | $upstream_response_time: [1.406] |
```
According to this log, request_time is often much larger than upstream_response_time. In fact, we do nothing but forward, apply a cross-domain plug-in to cros, and nothing else.

I don't understand why it takes so much time, sometimes even ten seconds extra.

Take this line:
```bash
 $request_time: [32.343] | $upstream_response_time: [23.665]
```
### 2.1 Pass postman authentication

I have done the verification, through kong and direct request file service, the conclusion is:

When passing kong, it really adds a lot of time.

This is a screenshot of my verification process, and the size of the picture is about 4M (the request parameters are the same).
![image](https://user-images.githubusercontent.com/49218070/230380067-3dee5f4c-55f2-439c-ac14-af0bd09a9fc3.png)
![image](https://user-images.githubusercontent.com/49218070/230380102-e3cdbeb4-d42f-4d31-9061-03a4f298ae9f.png)

### 2.2 Using wireshark to Capture Packets

It is the same process as above 2.1, except I carry out packet capture when passing kong.

My local ip is' 10.16.94.201 'and kong's address is' 10.7.133.10'.

I sorted the time spent in the request process and found that the time spent was on these nodes (looks like TCP timeout retransmission).
![image](https://user-images.githubusercontent.com/49218070/230380178-5a37e868-e543-4a12-bfb7-9757a343f89f.png)

I have tried to modify these two configuration items, but it seems that the problem is not there and the problem has not been solved.
![image](https://user-images.githubusercontent.com/49218070/230380256-ea050e1f-becb-4742-944d-3778e11a954e.png)

## 3. Questions

Now I don't know how to solve this problem, if necessary, I can also send my wireshark data file, I hope you can guide me, thank you.

**Comment by sabertobihwy** [comment]
First of all, file upload is a special scenario, and the default configuration of Kong does not have special optimization for file upload.

Below are some useful optimizations for file upload scenarios:

- `client_body_buffer_size`: This parameter sets the size of the buffer used to read the request body. A larger buffer can reduce disk I/O and improve performance. Adjust this value based on your needs and server memory. 

- `client_body_timeout`: This parameter sets the timeout for reading the request body. If the client connection speed is slow, you may need to increase this value to allow enough time for file uploads.

- `client_body_temp_path`: This parameter sets the path for storing uploaded files temporarily. Setting it to a high-speed disk or SSD can improve performance.

- `sendfile`: Enabling the sendfile option can use the operating system's sendfile() function to transfer data between disk and network, thereby improving performance. 

- `keepalive_timeout`: This parameter sets the timeout for long connections. Using long connections can reduce the time required to establish new connections, thereby improving performance. Adjust this value based on your needs. 

In short, I don't think this problem is related to Kong. You can refer to the optimization approach of Nginx for file upload scenarios.

**Comment by 1798972** [comment]
@sabertobihwy
Thank you very much for your reply. kong is a very useful gateway, and I also like it in my daily use.
Well, what puzzles me now is:
1. Parameter: Why is the difference between $upstream_response_time and $request_time so large?
2. Phenomenon: For the same picture (requested content), there is obvious difference between directly requesting the background picture service and requesting the background service through kong. `2.1 Pass postman authentication` has pictures.

**Comment by sabertobihwy** [comment]
I think you should try some cases:

Requests go through the Kong proxy without using any plugins, just a simple body. This case will show the time consumed by network.
Upload images through the Kong proxy without using any plugins. This case will show the time consumed by Kong when handling image proxy.
Based on this, we can analyze whether the specific delay is in the network, Kong handling images, or plugins like CORS.

**Comment by bungle** [comment]
If this is because Kong is buffereing data, you should use `route.enable_request_buffering=false`

**Comment by 1798972** [comment]
@bungle 
---
I verified that I set the buffer switch of the request and response to `flase`.
As route can see in the figure below, the status in `/ocrIdCardAndUploadOss` is already `false`.

![827e5f19975dfa36e50facc3a166b75a](https://user-images.githubusercontent.com/49218070/233597560-5a14702e-0cc2-4568-9ff8-2ccbc46c8f8f.png)
---
Then, I initiated the visit, and I can see that it has been successfully implemented.

![image](https://user-images.githubusercontent.com/49218070/233597503-9da76692-ec25-4b75-85d8-4940901a56d6.png)
---
Observing the log, I found that this phenomenon has not disappeared.
I tried several times, but the results were the same, and there was still a big time difference.
> kongDealTime: $request_time
 upRespTime: $upstream_response_time

![image](https://user-images.githubusercontent.com/49218070/233598605-44f2f28b-b5ec-4dd9-9588-a8d8c5143e5b.png)
---
My kong version is `2.2.1`.

![image](https://user-images.githubusercontent.com/49218070/233599109-b493b460-7948-4647-a091-e76e73d3a130.png)
---
Thanks again...

**Comment by 1798972** [comment]
@bungle 
---
I verified that I set the buffer switch of the request and response to `flase`.
As route can see in the figure below, the status in `/ocrIdCardAndUploadOss` is already `false`.

![827e5f19975dfa36e50facc3a166b75a](https://user-images.githubusercontent.com/49218070/233597560-5a14702e-0cc2-4568-9ff8-2ccbc46c8f8f.png)
---
Then, I initiated the visit, and I can see that it has been successfully implemented.

![image](https://user-images.githubusercontent.com/49218070/233597503-9da76692-ec25-4b75-85d8-4940901a56d6.png)
---
Observing the log, I found that this phenomenon has not disappeared.
I tried several times, but the results were the same, and there was still a big time difference.
> kongDealTime: $request_time
 upRespTime: $upstream_response_time

![image](https://user-images.githubusercontent.com/49218070/233598605-44f2f28b-b5ec-4dd9-9588-a8d8c5143e5b.png)
---
My kong version is `2.2.1`.

![image](https://user-images.githubusercontent.com/49218070/233599109-b493b460-7948-4647-a091-e76e73d3a130.png)
---
Thanks again...

**Comment by 1798972** [comment]
@greut 
By the way, there is only one acl plug-in on this route. I compared opening and closing the plug-in and found no significant difference in time.
Meaning, this is true even if I don't use any plugins.

**Comment by 1798972** [comment]
I verified that I set the buffer switch of the request and response to flase.
As route can see in the figure below, the status in /ocrIdCardAndUploadOss is already false.

827e5f19975dfa36e50facc3a166b75a
Then, I initiated the visit, and I can see that it has been successfully implemented.

image
Observing the log, I found that this phenomenon has not disappeared.
I tried several times, but the results were the same, and there was still a big time difference.

kongDealTime: $request_time
upRespTime: $upstream_response_time

image
My kong version is 2.2.1.

image
Thanks again...@bungle 
---
I verified that I set the buffer switch of the request and response to `flase`.
As route can see in the figure below, the status in `/ocrIdCardAndUploadOss` is already `false`.

![827e5f19975dfa36e50facc3a166b75a](https://user-images.githubusercontent.com/49218070/233597560-5a14702e-0cc2-4568-9ff8-2ccbc46c8f8f.png)
---
Then, I initiated the visit, and I can see that it has been successfully implemented.

![image](https://user-images.githubusercontent.com/49218070/233597503-9da76692-ec25-4b75-85d8-4940901a56d6.png)
---
Observing the log, I found that this phenomenon has not disappeared.
I tried several times, but the results were the same, and there was still a big time difference.
> kongDealTime: $request_time
 upRespTime: $upstream_response_time

![image](https://user-images.githubusercontent.com/49218070/233598605-44f2f28b-b5ec-4dd9-9588-a8d8c5143e5b.png)
---
My kong version is `2.2.1`.

![image](https://user-images.githubusercontent.com/49218070/233599109-b493b460-7948-4647-a091-e76e73d3a130.png)
---
By the way, there is only one acl plug-in on this route. I compared opening and closing the plug-in and found no significant difference in time.
Meaning, this is true even if I don't use any plugins.

---
Thanks again...

**Comment by sabertobihwy** [comment]
> I tried several times, but the results were the same, and there was still a big time difference.

but now `$request_time - $upstream_response_time`(time for Kong proxy) stable at around 1 second?

**Comment by sabertobihwy** [comment]
Can you use cURL instead of Postman on the machine where Kong is located (10.7.133.10) to send a request to the upstream service (10.7.12.130) and see the network latency and the time it takes to complete the request?

**Comment by 1798972** [comment]
@ghost
Thank you for your advice.

This is not the case. I tested a picture with a size of 4M.
![image](https://user-images.githubusercontent.com/49218070/236086163-92980c01-e346-45ce-beb3-dbeda726840f.png)
And all plug-ins have been closed.
![image](https://user-images.githubusercontent.com/49218070/236086260-ddee503d-2163-4439-be60-33bc54d7b50a.png)
There is still a big time difference.
![image](https://user-images.githubusercontent.com/49218070/236086443-2b88f28b-0ddb-456f-9365-89bff63cb939.png)

**Comment by 1798972** [comment]
@[bungle](https://github.com/bungle)
![image](https://github.com/Kong/kong/assets/49218070/321ce18c-38d5-44f6-bcb0-3f8bee24490b)

---

Just now I tried to update the buffer configuration again and the problem was fixed, but this is a new environment.
I suspect there are other reasons why the buffer switch issue hasn't been resolved for older environments, but the buffer switch should work.
thanks all~

---

![image](https://github.com/Kong/kong/assets/49218070/4ac7411a-08ad-4c6a-a4f7-654f9911f763)
